Modules/_abc.c

```diff
@@ -0,0 +1,847 @@
+    Py_DECREF(wr);
+    Py_DECREF(ref);
+    return ret;
+}
+
+/*[clinic input]
+_abc._reset_registry
+
+    self: object
+    /
+
+Internal ABC helper to reset registry of a given class.
+
+Should be only used by refleak.py
+[clinic start generated code]*/
+
+static PyObject *
+_abc__reset_registry(PyObject *module, PyObject *self)
+/*[clinic end generated code: output=92d591a43566cc10 input=12a0b7eb339ac35c]*/
+{
+    _abc_data *impl = _get_impl(self);
+    if (impl == NULL) {
+        return NULL;
+    }
+    if (impl->_abc_registry != NULL && PySet_Clear(impl->_abc_registry) < 0) {
+        Py_DECREF(impl);
+        return NULL;
+    }
+    Py_DECREF(impl);
+    Py_RETURN_NONE;
+}
+
+/*[clinic input]
+_abc._reset_caches
+
+    self: object
+    /
+
+Internal ABC helper to reset both caches of a given class.
+
+Should be only used by refleak.py
+[clinic start generated code]*/
+
+static PyObject *
+_abc__reset_caches(PyObject *module, PyObject *self)
+/*[clinic end generated code: output=f296f0d5c513f80c input=c0ac616fd8acfb6f]*/
+{
+    _abc_data *impl = _get_impl(self);
+    if (impl == NULL) {
+        return NULL;
+    }
+    if (impl->_abc_cache != NULL && PySet_Clear(impl->_abc_cache) < 0) {
+        Py_DECREF(impl);
+        return NULL;
+    }
+    /* also the second cache */
+    if (impl->_abc_negative_cache != NULL &&
+            PySet_Clear(impl->_abc_negative_cache) < 0) {
+        Py_DECREF(impl);
+        return NULL;
+    }
+    Py_DECREF(impl);
+    Py_RETURN_NONE;
+}
+
+/*[clinic input]
+_abc._get_dump
+
+    self: object
+    /
+
+Internal ABC helper for cache and registry debugging.
+
+Return shallow copies of registry, of both caches, and
+negative cache version. Don't call this function directly,
+instead use ABC._dump_registry() for a nice repr.
+[clinic start generated code]*/
+
+static PyObject *
+_abc__get_dump(PyObject *module, PyObject *self)
+/*[clinic end generated code: output=9d9569a8e2c1c443 input=2c5deb1bfe9e3c79]*/
+{
+    PyObject *registry, *cache, *negative_cache, *cache_version;
+    _abc_data *impl = _get_impl(self);
+    if (impl == NULL) {
+        return NULL;
+    }
+    registry = PySet_New(impl->_abc_registry);
+    if (registry == NULL) {
+        Py_DECREF(impl);
+        return NULL;
+    }
+    cache = PySet_New(impl->_abc_cache);
+    if (cache == NULL) {
+        Py_DECREF(impl);
+        Py_DECREF(registry);
+        return NULL;
+    }
+    negative_cache = PySet_New(impl->_abc_negative_cache);
+    if (negative_cache == NULL) {
+        Py_DECREF(impl);
+        Py_DECREF(registry);
+        Py_DECREF(cache);
+        return NULL;
+    }
+    cache_version = PyLong_FromUnsignedLongLong(impl->_abc_negative_cache_version);
+    if (cache_version == NULL) {
+        Py_DECREF(impl);
+        Py_DECREF(registry);
+        Py_DECREF(cache);
+        Py_DECREF(negative_cache);
+        return NULL;
+    }
+    PyObject *res = PyTuple_Pack(4, registry, cache, negative_cache, cache_version);
```

Comment:
This could be written simpler as:
```c
PyObject *res = Py_BuildValue("NNNK",
                              PySet_New(impl->_abc_registry),
                              PySet_New(impl->_abc_cache),
                              PySet_New(impl->_abc_negative_cache),
                              impl->_abc_negative_cache_version);
```